### PR TITLE
In the semantics tree, do not detach a child if it has already been assigned a new parent

### LINKED ISSUE #6690: Flash of red exception on App Bar when returning to home screen in Gallery
I've seen this on iOS hardware and simulator. Have not tested on Android.

open gallery, go into Shrine, then tap back.

Notice the flash of red exception where the app bar is. The app bar will replace the red exception quickly.

### PR DESCRIPTION
**Pull request by jason-simmons**
Fixes https://github.com/flutter/flutter/issues/6690

Also add a version of the Gallery smoke test that enables semantics

### CODE REVIEW

**Comment by jason-simmons** [comment]
@Hixie

File: examples/flutter_gallery/test/smoke_test_semantics.dart [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// found in the LICENSE file.
+
+import 'package:flutter/rendering.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+import 'smoke_test.dart';
+
+void main() {
+  testWidgets('Enable semantics', (WidgetTester tester) async {
+    RendererBinding.instance.setSemanticsEnabled(true);
```
Comment (Hixie) [inline examples/flutter_gallery/test/smoke_test_semantics.dart]: in theory nothing guarantees the order the tests run in and they shouldn't depend on each other.

I think the right way to factor this out would be to have the inside of the "Flutter Gallery app smoke test" in its own function, and then have the two tests both call that function. The semantics one should enable the semantics before and disable them after. I think it'd be fine for these two tests to be in the same file (otherwise it's a bit weird because one test imports the other).

File: packages/flutter/lib/src/rendering/semantics.dart [hunk 2]
```diff
@@ -473,8 +474,11 @@ class SemanticsNode extends AbstractNode {
     owner._detachedNodes.add(this);
     super.detach();
     if (_children != null) {
-      for (SemanticsNode child in _children)
-        child.detach();
+      for (SemanticsNode child in _children) {
+        if (child.parent == this) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/semantics.dart]: add a comment saying why this might not be true

**Comment by Hixie** [comment]
LGTM

File: examples/flutter_gallery/test/smoke_test_semantics.dart (hunk 1 above)
Comment (jason-simmons) [inline examples/flutter_gallery/test/smoke_test_semantics.dart]: done

File: examples/flutter_gallery/test/smoke_test.dart [hunk 3]
```diff
@@ -47,10 +48,8 @@ Future<Null> smokeDemo(WidgetTester tester, String routeName) async {
   return null;
 }
 
-void main() {
-  testWidgets('Flutter Gallery app smoke test', (WidgetTester tester) async {
-    flutter_gallery_main
-        .main(); // builds the app and schedules a frame but doesn't trigger one
+Future<Null> runSmokeTest(WidgetTester tester) async {
+    await tester.pumpWidget(new GalleryApp());
```
Comment (Hixie) [inline examples/flutter_gallery/test/smoke_test.dart]: nit: this is indented 4 instead of 2

File: packages/flutter/lib/src/rendering/semantics.dart [hunk 4]
```diff
@@ -473,8 +474,13 @@ class SemanticsNode extends AbstractNode {
     owner._detachedNodes.add(this);
     super.detach();
     if (_children != null) {
-      for (SemanticsNode child in _children)
-        child.detach();
+      for (SemanticsNode child in _children) {
+        // The list of children may be stale and may contain nodes that have
+        // been assigned to a different parent.
+        if (child.parent == this) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/semantics.dart]: nit: extraneous braces

**Comment by Hixie** [comment]
still LGTM
